jax/_src/pjit.py

```diff
@@ -475,21 +478,31 @@ def common_infer_params(pjit_info_args, *args, **kwargs):
 
   del user_in_shardings, user_out_shardings
 
-  global_in_avals = tuple(shaped_abstractify(a) for a in args_flat)
+  if config.jax_dynamic_shapes:
+    in_type = pe.infer_lambda_input_type(axes_specs, explicit_args)
+  else:
+    in_type = tuple(shaped_abstractify(a) for a in explicit_args)
 
   canonicalized_in_shardings_flat = _process_in_axis_resources(
-      hashable_pytree(in_shardings), global_in_avals, in_tree, resource_env)
+      hashable_pytree(in_shardings), in_type, in_tree, resource_env)
 
   jaxpr, consts, canonicalized_out_shardings_flat = _pjit_jaxpr(
-      flat_fun, hashable_pytree(out_shardings), global_in_avals, dbg,
+      flat_fun, hashable_pytree(out_shardings), in_type, dbg,
       HashableFunction(out_tree, closure=()),
       HashableFunction(res_paths, closure=()))
 
   if any(_is_from_gda(i) for i in canonicalized_in_shardings_flat):
+    if config.jax_dynamic_shapes: raise NotImplementedError
     canonicalized_in_shardings_flat = _maybe_replace_from_gda_with_pspec(
-        canonicalized_in_shardings_flat, args_flat)
+        canonicalized_in_shardings_flat, explicit_args)
 
+  if config.jax_dynamic_shapes:
+    implicit_args = _extract_implicit_args(in_type, explicit_args)
+  else:
+    implicit_args = []
+  args_flat = [*implicit_args, *explicit_args]
   assert len(args_flat) == len(canonicalized_in_shardings_flat)
+  donated_invars = (False,) * len(implicit_args) + donated_invars
```

Comment:
I wonder if this can be merged below on line 509?

Also if implicit_args are present, you need to handle canonicalized_in_shardings_flat too? Which can also be merged on line 507?

donated_invars = (False,) * (len(implicit_args) + len(consts)) + donated_invars

Reply:
Done!